### PR TITLE
[Relay] add Tuple pattern

### PR DESCRIPTION
**Pull request by MarisaKirisame**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).
@icemelon9 @slyubomirsky @Yulun-Yao @vinx13 can you guys help review?

### CODE REVIEW

File: src/relay/backend/interpreter.cc [hunk 1]
```diff
@@ -18,7 +18,7 @@
  */
 
 /*!
- *  Copyright (c) 2018 by Contributors
+ *  Copyright (c) 2019 by Contributors
```
Comment (u99127) [inline src/relay/backend/interpreter.cc]: In my experience such changes are usually a year range. does the ASF not need a year range here ?
git blame suggests this file was created in 2018, thus this should read copyright 2018-19 IIUC. However IANAL and the exact definition depends on project policy.

@tqchen , any thoughts

File: src/relay/ir/pattern_functor.cc [hunk 2]
```diff
@@ -18,8 +18,8 @@
  */
 
 /*!
- *  Copyright (c) 2018 by Contributors
- * \file src/tvm/relay/pattern_functor.cc
+ *  Copyright (c) 2019 by Contributors
```
Comment (u99127) [inline src/relay/ir/pattern_functor.cc]: I think this is a typo - however @tqchen needs to comment as he was the last one who changed this. The first commit came in 2019, so you are safe here.

File: src/relay/backend/interpreter.cc (hunk 1 above)
Comment (MarisaKirisame) [reply src/relay/backend/interpreter.cc]: Thank you for catching this! I just wrongly assume to use the latest year whenever I update a file.

**Comment by slyubomirsky** [comment]
Very glad you added this; it will certainly be a convenience. I will give a detailed review later.

**Comment by slyubomirsky** [comment]
High-level feedback that you probably guessed: We'll need to add test cases for the features where you added tuple patterns (type inference, the interpreter, match completeness checking, partial eval, etc.)

**Comment by MarisaKirisame** [comment]
@slyubomirsky I changed the prelude zip, so it work for ti/interpreter/vm/match completeness without any additional code. Partial Eval also PE all prelude function so at least the code run and typecheck.

**Comment by jroesch** [comment]
@slyubomirsky can you explicitly review cc @wweic

**Comment by slyubomirsky** [comment]
Again, I think _specific_ unit tests (for alpha-equality, match completeness, etc.), not just by changing the implementations of prelude functions (though that's handy too), would still be very important to have.

I have read over the implementation of the feature and do not have any issues with it, but I stress the importance of having separate tests since if something goes wrong or a later feature interacts badly with what we have, specific tests will make it much easier to identify what went wrong.

**Comment by jroesch** [comment]
@MarisaKirisame and @slyubomirsky can you guys get this one over the finish line this week?

**Comment by MarisaKirisame** [comment]
@jroesch @slyubomirsky I had added the test. it is now good.

**Comment by MarisaKirisame** [comment]
@vinx13 @slyubomirsky @wweic @junrushao1994 can you guys give another round of review?

**Comment by slyubomirsky** [comment]
Thanks for adding tests. LGTM

**Pull request merged**